### PR TITLE
cephadm: truncate X.509 Common Name to 64 chars in generate_cert

### PR DESCRIPTION
**Pull request by joshjan20**
**Scope note (added after review):** this PR hardens generate_cert() defensively against a long addrs[0], but does NOT fix the originally reported issue (#80277). Tracing the actual call chain on squid 19.2.6 (where the bug was reported) shows Grafana's cert generation goes through a completely different path, monitoring.py's prepare_certificates() calling mgr_util.create_self_signed_cert() directly, which has no length check at all. The real fix belongs there; I'll open a separate PR for it. This PR remains worthwhile on its own: main's ssl_certs.generate_cert() would still fail the same way if any future caller ever passes a long hostname as addrs[0].

---

The Common Name (CN) field in the certificate subject is limited to 64 characters per RFC 5280. generate_cert() previously passed the raw address/hostname directly as the CN with no length check. On cloud providers with long auto-assigned FQDNs (e.g. Google Cloud Platform, which embeds the project ID into every VM hostname), this can easily exceed 64 characters, causing certificate signing to fail with an opaque low-level error:

  asn1 encoding routines: ... string too long

This surfaces to the user as an unexplained failure to deploy services that require a generated certificate (e.g. Grafana), with no indication that hostname length is the actual cause.

TLS hostname verification relies on the SAN (Subject Alternative Name) list, not the CN. The full, untruncated hostname is already present in the SAN, added separately in this same method. Truncating only the CN when it exceeds 64 characters therefore has no effect on certificate validity or hostname verification; it only avoids the signing failure.

The fix truncates the CN at the first DNS label when needed, rather than a blind character slice, to avoid cutting the CN mid-word. This does not affect certificate validity or TLS hostname verification, which relies on the SAN (Subject Alternative Name) list, not the CN; the full, untruncated hostname remains present in the SAN, unmodified.

Includes 5 regression tests, verified passing locally in a standalone environment (see commit message for details); full test suite integration will run via CI once labeled.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

### CODE REVIEW

**Comment by joshjan20** [comment]
This is now tracked upstream: https://tracker.ceph.com/issues/80277

File: src/pybind/mgr/cephadm/ssl_certs.py [hunk 1]
```diff
@@ -93,7 +103,18 @@ def generate_cert(
...
+        # providers (notably GCP) can assign FQDNs well past that limit, and
+        # exceeding it fails certificate signing with an opaque low-level
+        # error rather than a clear one. TLS hostname verification relies on
+        # the SAN list (populated below from `hosts`), not the CN, so
+        # truncating the CN here does not affect the certificate's actual
+        # validity for the hosts it covers; the full hostname is still
+        # present, unmodified, in the SAN.
+        cn_value = addrs[0]
+        if len(cn_value) > COMMON_NAME_MAX_LENGTH:
+            cn_value = cn_value[:COMMON_NAME_MAX_LENGTH]
```
Comment (tchaikov) [inline src/pybind/mgr/cephadm/ssl_certs.py:116]: on main and tentacle, since 8be392f62b1, grafana's cert is signed by cephadm with the host's inventory address as the CN and the FQDN in the SAN, so a long FQDN alone does not break it. i tried the unpatched `generate_cert()` with the hostname from the ticket and an ip as `addrs`, and it works. so what https://tracker.ceph.com/issues/80277 needs is a squid fix in `prepare_certificates()`.

File: src/pybind/mgr/cephadm/tests/test_ssl_certs.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    def test_long_fqdn_does_not_raise(self):
+        """A CN over 64 characters must not raise during certificate generation."""
+        long_fqdn = "ceph-node1.europe-west3-a.c.project-12850071-31c6-4077-a2f.internal"
+        self.assertGreater(len(long_fqdn), COMMON_NAME_MAX_LENGTH)
+
+        # Should not raise ValueError: Attribute's length must be >= 1 and <= 64
+        cert_pem, key_pem = self.certs.generate_cert(
+            _hosts=[long_fqdn],
+            _addrs=[long_fqdn],
```
Comment (tchaikov) [inline src/pybind/mgr/cephadm/tests/test_ssl_certs.py:31]: `generate_cert()` never adds a non-ip entry of `addrs` to the SAN, so the full name is only there if the caller also passes it in `hosts`. grafana does, but the method does not guarantee it. the tests pass the same string as both `hosts` and `addrs`, so they would not notice. prbly we can test with what grafana passes: `hosts=[fqdn, 'grafana_servers']` and `addrs=[fqdn]`?

File: src/pybind/mgr/cephadm/ssl_certs.py [hunk 3]
```diff
@@ -93,7 +103,18 @@ def generate_cert(
...
-        builder = builder.subject_name(x509.Name([x509.NameAttribute(NameOID.COMMON_NAME, addrs[0]), ]))
+        # X.509 Common Name is limited to 64 characters (RFC 5280). Cloud
+        # providers (notably GCP) can assign FQDNs well past that limit, and
+        # exceeding it fails certificate signing with an opaque low-level
+        # error rather than a clear one. TLS hostname verification relies on
+        # the SAN list (populated below from `hosts`), not the CN, so
+        # truncating the CN here does not affect the certificate's actual
+        # validity for the hosts it covers; the full hostname is still
+        # present, unmodified, in the SAN.
+        cn_value = addrs[0]
```
Comment (tchaikov) [inline src/pybind/mgr/cephadm/ssl_certs.py:114]: instead of cutting the name at 64 chars, which leaves something like `...a2f.inter`, how about using the first label when it is too long? a dns label is at most 63 chars, so it always fits, and ips are never that long, so they are not affected:
  ```python
  cn_value = addrs[0]
  if len(cn_value) > COMMON_NAME_MAX_LENGTH:
      cn_value = cn_value.split('.')[0]
```

File: src/pybind/mgr/cephadm/ssl_certs.py [hunk 4]
```diff
@@ -9,6 +9,16 @@
...
 from ceph.deployment.tls_utils import SSLConfigException
 
+# RFC 5280 defines the X.509 Common Name (CN) field as a DirectoryString,
+# whose underlying ASN.1 types (PrintableString, UTF8String, etc.) impose a
+# maximum length of 64 characters. Passing a longer value causes certificate
+# signing to fail with an opaque low-level error (OpenSSL's
+# "asn1 encoding routines: ... string too long"), which gives no indication
+# that the CN length is the actual cause. This is most commonly hit with
+# auto-assigned cloud FQDNs (e.g. Google Cloud Platform embeds the project
+# ID into the hostname, easily producing FQDNs well over 64 characters).
```
Comment (tchaikov) [inline src/pybind/mgr/cephadm/ssl_certs.py:19]: the rationale is repeated here *and* in the test docstring, probly we can keep it in the commit message only.

File: src/pybind/mgr/cephadm/ssl_certs.py (hunk 1 above)
Comment (joshjan20) [reply src/pybind/mgr/cephadm/ssl_certs.py]: You're right, and I traced it through to confirm. On squid, the Grafana cert path never goes through ssl_certs.generate_cert() at all: monitoring.py's GrafanaService.prepare_certificates() calls mgr_util.create_self_signed_cert('Ceph', host_fqdn) directly, which uses pyOpenSSL and has no length check anywhere. This is genuinely where the original failure came from on squid 19.2.6, which is reported in #80277.

So you're correct: this PR is defensive hardening for ssl_certs.generate_cert(), worth keeping on its own merits since a long addrs[0] would still break it if any future caller passes one, but it doesn't fix #80277. Updated the PR description to say so explicitly, and I'll open a separate PR against the actual code path.

Thanks for pushing on this, it would've shipped as a fix that doesn't fix the reported bug.

File: src/pybind/mgr/cephadm/tests/test_ssl_certs.py (hunk 2 above)
Comment (joshjan20) [reply src/pybind/mgr/cephadm/tests/test_ssl_certs.py:22]: Updated in the latest commit, tests now use distinct _hosts/_addrs values (hosts=[fqdn, 'grafana_servers'], addrs=[fqdn]), so the SAN assertions actually exercise the split.

File: src/pybind/mgr/cephadm/ssl_certs.py (hunk 3 above)
Comment (joshjan20) [reply src/pybind/mgr/cephadm/ssl_certs.py:101]: Switched to cn_value.split('.')[0][:COMMON_NAME_MAX_LENGTH], truncating at the first DNS label with a character-slice fallback for the dot-less edge case you flagged. Added a new test covering that edge case, and updated the old test that used a dot-less string (it would've broken under the new logic).

File: src/pybind/mgr/cephadm/ssl_certs.py (hunk 4 above)
Comment (joshjan20) [reply src/pybind/mgr/cephadm/ssl_certs.py]: Trimmed both the module comment and the code comment down to one/two lines each, consolidated what had become duplicate text in the process. Full rationale stays in the commit message.
